### ISSUE TITLE
GraphQL API  can't get the nested details and can't get the proper content when created a custom graphql

### ISSUE THREAD

**Issue by fr33css** [aws-amplify/amplify-swift#2806]
### Describe the bug

xcode 14.2
amplify 10.8.1

with this schema
```
type User @model @auth(rules: [{allow: public}]) {
  id: ID!
  username: String!
  posts: [Post] @hasMany(indexName: "byUser", fields: ["id"])
  comments: [Comment] @hasMany(indexName: "byUser", fields: ["id"])
  threads: [Thread] @hasMany(indexName: "byUser", fields: ["id"])
}

type Thread @model @auth(rules: [{allow: public}]) {
  id: ID!
  commentID: ID! @index(name: "byComment")
  content: String!
  userID: ID! @index(name: "byUser")
}

type Comment @model @auth(rules: [{allow: public}]) {
  id: ID!
  content: String!
  postID: ID! @index(name: "byPost", sortKeyFields: ["content"])
  post: Post! @belongsTo(fields: ["postID"])
  threads: [Thread] @hasMany(indexName: "byComment", fields: ["id"])
  userID: ID! @index(name: "byUser")
}

type Post @model @auth(rules: [{allow: public}]) {
  id: ID!
  title: String!
  description: String
  comments: [Comment] @hasMany(indexName: "byPost", fields: ["id"])
  userID: ID! @index(name: "byUser")
}
 
```
### Issue 1

when doing

```
let request = GraphQLRequest<Post>.list(Post.self)

try await Amplify.API.query(request: request)
```

i can't to get the comments details

```
Post(id: "FC51010E-517E-40F4-AE59-A98317298348", title: "Post 1", description: nil, comments: Optional(Amplify.List<MyFirstApp.Comment>), userID: "c273f24c-7c22-42bc-9f91-102b42646347", createdAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2023-03-09 08:48:25 +0000)), updatedAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2023-03-09 08:48:25 +0000))), MyFirstApp.Post(id: "0ED9558E-A4D8-447E-BF93-1047F196BE66", title: "Post 2", description: Optional("description"), comments: Optional(Amplify.List<MyFirstApp.Comment>), userID: "c273f24c-7c22-42bc-9f91-102b42646347", createdAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2023-03-09 10:25:51 +0000)), updatedAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2023-03-09 10:25:51 +0000)))]
```

###  Issue 2
Follow the link https://docs.amplify.aws/lib/graphqlapi/advanced-workflows/q/platform/ios/#nested-data

Create own GraphQLRequest
```
extension GraphQLRequest {

    static func getPostId(byId id: String) -> GraphQLRequest<Post> {
        let document = """
        query getPost($id: ID!) {
          getPost(id: $id) {
            id
          }
        }
        """
        return GraphQLRequest<Post>(
            document: document,
            variables: ["id": id],
            responseType: Post.self,
            decodePath: "getPost"
        )
    }
}
```

when doing

`try await Amplify.API.query(request: .getPostId(byId: id))`

then return error

```
Failed to create graphql GraphQLResponseError<Post>: Failed to decode GraphQL response to the `ResponseType` Post
Recovery suggestion: Failed to transform to `ResponseType`.
Take a look at the `RawGraphQLResponse` and underlying error to see where it failed to decode.
Caused by:
APIError: keyNotFound key CodingKeys(stringValue: "title", intValue: nil)
Caused by:
keyNotFound(CodingKeys(stringValue: "title", intValue: nil), Swift.DecodingError.Context(codingPath: [], debugDescription: "No value associated with key CodingKeys(stringValue: \"title\", intValue: nil) (\"title\").", underlyingError: nil))
```

When i run at the AWS AppSync Queries Console
i can get back the nested detail and created my custom graphql
<img width="500" alt="Screenshot 2023-03-09 at 20 51 01" src="https://user-images.githubusercontent.com/56120515/224015001-26db7dd0-f11a-4a22-85f4-24ec01f3fc79.png">

### Steps To Reproduce

```swift
amplify init
amplify add auth (cognito)
amplify add api (with conflict detection-> auto merge)
amplify add storage
amplify codegen models
amplify push
```

### Expected behavior

Issue 1 

return the nested detail(comments details)

Issue 2

no errors

### Amplify Framework Version

2.0

### Amplify Categories

API

### Dependency manager

Swift PM

### Swift version

5.0

### CLI version

10.8.1

### Xcode version

14.2

### Relevant log output

```shell
For Issue 1

Post(id: "FC51010E-517E-40F4-AE59-A98317298348", title: "Post 1", description: nil, comments: Optional(Amplify.List<MyFirstApp.Comment>), userID: "c273f24c-7c22-42bc-9f91-102b42646347", createdAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2023-03-09 08:48:25 +0000)), updatedAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2023-03-09 08:48:25 +0000))), MyFirstApp.Post(id: "0ED9558E-A4D8-447E-BF93-1047F196BE66", title: "Post 2", description: Optional("description"), comments: Optional(Amplify.List<MyFirstApp.Comment>), userID: "c273f24c-7c22-42bc-9f91-102b42646347", createdAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2023-03-09 10:25:51 +0000)), updatedAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2023-03-09 10:25:51 +0000)))]




For Issue 2

Failed to create graphql GraphQLResponseError<Post>: Failed to decode GraphQL response to the `ResponseType` Post
Recovery suggestion: Failed to transform to `ResponseType`.
Take a look at the `RawGraphQLResponse` and underlying error to see where it failed to decode.
Caused by:
APIError: keyNotFound key CodingKeys(stringValue: "title", intValue: nil)
Caused by:
keyNotFound(CodingKeys(stringValue: "title", intValue: nil), Swift.DecodingError.Context(codingPath: [], debugDescription: "No value associated with key CodingKeys(stringValue: \"title\", intValue: nil) (\"title\").", underlyingError: nil))
```

### Is this a regression?

No

### Device

iPhone 14 Pro Max simulator

### iOS Version

16.2

**Comment by royjit** [comment]
Hi @fr33css Thank you for reaching out,

Comments are lazy loaded, to get the details you need to fetch them first:
```swift
           guard let comments = post.comments else {
                print("No comments received")
                return
            }
            try await comments.fetch()
            for comment in comments {
                print(comment)
            }
```

Issue 2

We need to pass all the required parameters while you create the request, `title` and `userID` were missing from the code you provided:

```swift
extension GraphQLRequest {

    static func getPostId(byId id: String) -> GraphQLRequest<Post> {
        let document = """
        query getPost($id: ID!) {
          getPost(id: $id) {
            id
            title
            userID
          }
        }
        """
        return GraphQLRequest<Post>(
            document: document,
            variables: ["id": id],
            responseType: Post.self,
            decodePath: "getPost"
        )
    }
}
```

Let us know if the above changes unblocks you.

**Comment by fr33css** [comment]
Hi @royjit, thanks for you response.

According Issue 1, I followed your suggestions.

| AWS Appsync Console | Xcode Log | 
| -- | -- |
| <img width="1026" alt="Screenshot 2023-03-12 at 13 20 20" src="https://user-images.githubusercontent.com/56120515/224524169-012f1134-7c88-4019-8d1e-52afa6b74b2e.png"> | <img width="684" alt="Screenshot 2023-03-12 at 13 18 43" src="https://user-images.githubusercontent.com/56120515/224524270-703e8a93-5de2-4f23-821e-7417014be4d6.png"> |

```
func getPost(byId id: String) async {
     do {
         let result = try await Amplify.API.query(request: .getPostId(byId: id))
         switch result {
         case .success(let post):
             print(post)
             guard let comments = post.comments else {
                 print("No comments received")
                 return
             }
             try await comments.fetch()
             for comment in comments {
                 print(comment)
             }
         case .failure(let graphQLError):
             print("Failed to create graphql \(graphQLError)")
         }
     } catch let error as APIError {
         print("Failed to create a todo: ", error)
     } catch {
         print("Unexpected error: \(error)")
     }
}
```

But i still can't get any comments back. 

Issue 2 

> We need to pass all the required parameters while you create the request, title and userID were missing from the code you provided:

I see. 
so if we just want to return id, then should make the title and userID is not required?
because regarding the link https://docs.amplify.aws/lib/graphqlapi/advanced-workflows/q/platform/ios/#nested-data

the post model with id,title,rating,status is required.
```
type Post @model {
  id: ID!
  title: String!
  rating: Int!
  status: PostStatus!
  comments: [Comment] @hasMany(indexName: "byPost", fields: ["id"])
  notes: [Note] @hasMany(indexName: "byNote", fields: ["id"])
}
```

```
extension GraphQLRequest {
    static func get(byPostId postId: String, todoId: String) -> GraphQLRequest<JSONValue> {
        let document = """
        query get($postId: ID!, $todoId: ID!) {
          getPost(id: $postId) {
            id
            title
            rating
          }
          getTodo(id: $todoId) {
            id
            name
          }
        }
        """

        return GraphQLRequest<JSONValue>(
            document: document,
            variables: [
                "postId": postId,
                "todoId": todoId
            ],
            responseType: JSONValue.self
        )
    }
}
```

but we can get the Post by using getPost to return only id,title,rating. 
no need including the status which is required.

**Comment by royjit** [comment]
Looks like `__typename` is required for the custom graphql to work. Here is what I tried with your schema:

```swift
extension GraphQLRequest {

    static func getPostId(byId id: String) -> GraphQLRequest<Post> {
        let document = """
        query GetPost($id: ID!) {
          getPost(id: $id) {
            id
            title
            userID
            __typename
          }
        }
        """
        return GraphQLRequest<Post>(
            document: document,
            variables: ["id": id],
            responseType: Post.self,
            decodePath: "getPost"
        )
    }
}
```

Below query was able to print comments:

```swift
           let result = try await Amplify.API.query(request: .getPostId(byId: savedId))

            var comments: List<Comment>?
            switch result {
            case .success(let post):
                comments = post.comments
            case .failure(let error):
                print(error)
            }
            guard let ncomments = comments else {
                print("No comments received")
                return
            }
            try await ncomments.fetch()
            for comment in ncomments {
                print(comment)
            }
```

Please check if the above fixes your use case. I will check the documentation example at https://docs.amplify.aws/lib/graphqlapi/advanced-workflows/q/platform/ios/#nested-data to see if there is a typo.

**Comment by royjit** [comment]
Looks like the code example you took from the documentation is contains Post and Todo which is not related to the schema provided in the doc ie, `Todo` is not present in the schema provided. 

I tried to verify and found that the code does not work if required variables are not present in the selection set. I will create a new PR against doc repo to fix.

**Comment by fr33css** [comment]
@royjit 
<img width="800" alt="Screenshot 2023-03-14 at 10 36 57" src="https://user-images.githubusercontent.com/56120515/224869800-b1214dac-0d97-49d1-bee9-638a2aff4af3.png">

```
 let result = try await Amplify.API.query(request: .getPostId(byId: savedId))

            var comments: List<Comment>?
            switch result {
            case .success(let post):
                comments = post.comments
            case .failure(let error):
                print(error)
            }
            guard let ncomments = comments else {
                print("No comments received")
                return
            }
            try await ncomments.fetch()
            for comment in ncomments {
                print(comment)
            }

```

I followed your suggestions but having the List<Comment>? error problem. 
How can i fix it ?

**Comment by royjit** [comment]
If you open `Post.swift` file in your xcode what is the type for `comments`? Mine is this:

```swift
public var comments: List<Comment>?
```

I used the same to hold on to that variable. Could you share the Amplify Swift version and Amplify CLI version you are using, wondering if there is some type mismatch based on the version.

**Comment by fr33css** [comment]
i know what the problem is because if 
```
import Amplify and
import SwiftUI together
```
`public var comments: List<Comment>?`

will become error

**Comment by fr33css** [comment]
and one more question

https://docs.amplify.aws/lib/graphqlapi/mutate-data/q/platform/ios/

Regarding the link,
Can i know the proper way to delete post and update post ?

```
func deletePost(id: String, name: String, userId: String) async {
        let post = Post(id: id, title: name, userID: userId)
        do {
            let result = try await Amplify.API.mutate(request: .delete(post))
            switch result {
            case .success(let data):
                switch result {
                case .success(let post):
                    print("Successfully deleted post: \(post)")
                case .failure(let error):
                    print("Got failed result with \(error.errorDescription)")
                }
            case .failure(let graphQLError):
                print("Failed to create graphql \(graphQLError)")
            }
        } catch {
            print("Delete Blog - \(error.localizedDescription)")
        }
    }
```

It's getting error

```
Failed to create graphql GraphQLResponseError<Post>: GraphQL service returned a successful response containing errors: [Amplify.GraphQLError(message: "Conflict resolver rejects mutation.", locations: Optional([Amplify.GraphQLError.Location(line: 2, column: 3)]), path: Optional([Amplify.JSONValue.string("deletePost")]), extensions: Optional(["errorType": Amplify.JSONValue.string("ConflictUnhandled"), "data": Amplify.JSONValue.object(["updatedAt": Amplify.JSONValue.string("2023-03-15T03:37:25.151Z"), "createdAt": Amplify.JSONValue.string("2023-03-15T03:37:25.151Z"), "description": Amplify.JSONValue.string("description"), "__typename": Amplify.JSONValue.string("Post"), "title": Amplify.JSONValue.string("Post"), "id": Amplify.JSONValue.string("5FDD2870-3DC6-45FB-A3C9-8D451CF56E7C"), "userID": Amplify.JSONValue.string("1d13c716-0d1b-4bc8-9c5d-7a0ea50f73ce")]), "errorInfo": Amplify.JSONValue.null]))]
Recovery suggestion: The list of `GraphQLError` contains service-specific messages
```

**Comment by royjit** [comment]
Your code looks okay to me, I will try to replicate at my side. Do you have Datastore and API category added in your app, I [read here](https://github.com/aws-amplify/amplify-android/issues/591) that these kind of issue can happen if you use apis from `Amplify.API` and `Amplify.Datastore` together in the same app.

**Comment by royjit** [comment]
As mentioned here - https://docs.amplify.aws/lib/graphqlapi/getting-started/q/platform/ios/#prerequisites Amplify.API category doesnot support conflict resolution. If you need conflict resolution handling please use Amplify Datastore.

> Troubleshooting: The AWS API plugins do not support conflict detection. If AppSync returns errors about missing _version and _lastChangedAt fields, or unhandled conflicts, disable conflict detection. Run amplify update api, and choose Disable conflict detection. If you started with the Getting Started guide, which introduces DataStore, this step is required.

**Comment by fr33css** [comment]
@royjit 
Thank you.
Ya, i using Amplify API and Amplify.Datastore for uploading image to s3Bucket.
I can update successfully at first time.
After of that , start shown a new column [_version] and can't update the informations.
do you know how to fix it ?

<img width="155" alt="Screenshot 2023-03-19 at 21 04 48" src="https://user-images.githubusercontent.com/56120515/226173943-d8fc08bc-601e-4d37-9b6f-0329ec45b892.png">

**Comment by fr33css** [comment]
@royjit 

Thank you for your response.

https://github.com/aws-amplify/amplify-swift/issues/2806#issuecomment-1467073576
According for your customGraphQL:

i still get all the informations of the posts or comments.
As you said if is the primarykey if needed like id, title, userID from the post.

```
extension GraphQLRequest {

    static func getPostId(byId id: String) -> GraphQLRequest<Post> {
        let document = """
        query getPost($id: ID!) {
          getPost(id: $id) {
            id
            title
            userID
          }
        }
        """
        return GraphQLRequest<Post>(
            document: document,
            variables: ["id": id],
            responseType: Post.self,
            decodePath: "getPost"
        )
    }
}

```

but it's still getting the all of the informations of 
`comments,description,createdAt, updatedAt`
it have no meaning for the custom GraphQL

**Comment by royjit** [comment]
Sorry I do not understand your last comment, could you explain it bit more?

**Comment by fr33css** [comment]
@royjit sorry for bad english.

```
type User @model @auth(rules: [{allow: public}]) {
  id: ID!
  emailAddress: String!
  username: String!
  bio: String
}
```

for using a custom graphQL,
i just want to return (user response) id, emailAddress,username only.
i don't want another information like `bio,createdAt, updatedAt`...

extension GraphQLRequest {

    static func getUserInfo(byId id: String) -> GraphQLRequest<User> {
        let document = """
        query getUser($id: ID!) {
          getUser(id: $id) {
            id
            username
            emailAddress
          }
        }
        """
        return GraphQLRequest<User>(
            document: document,
            variables: ["id": id],
            responseType: User.self,
            decodePath: "getUser"
        )
    }
}

even though i made a custom GraphQL it's still getting full response.
Expected returning `id, username, emailAddress` 

but 
`id, emailAddress, username, bio, createdAt, updatedAt` is still showing out.

**Comment by royjit** [comment]
Thank you for explaining, I was not able to reproduce the issue. When I tried to query :

```swift
extension GraphQLRequest {
    static func getcustomPost(byId id: String) -> GraphQLRequest<Post> {
          let document = """
          query getPost($id: ID!) {
            getPost(id: $id) {
              id
              name
            }
          }
          """
          return GraphQLRequest<Post>(
              document: document,
              variables: ["id": id],
              responseType: Post.self,
              decodePath: "getPost"
          )
      }
}
```
On a schema:
```
type Blog @model @auth(rules: [{allow: public}]) {
  id: ID!
  name: String!
  post: [Post] @hasMany(indexName: "postByBlog", fields: ["id"])
}

type Post @model @auth(rules: [{allow: public}]) {
  id: ID!
  name: String!
  blog_id: ID @index(name: "postByBlog")
  blog: Blog @belongsTo(fields: ["blog_id"])
}

```
I did not get other fields. Here is verbose log when I invoke API request:

```
2023-03-28 16:55:00.894671-0700 DatastoreAPISample[87093:3235014] [API] Starting query 3D3F733D-5C32-4347-ABFF-7005324B4BB5
2023-03-28 16:55:00.894896-0700 DatastoreAPISample[87093:3235014] [API] {
  "query" : "query getPost($id: ID!) {\n  getPost(id: $id) {\n    id\n    name\n  }\n}",
  "variables" : {
    "id" : "895f4988-e7fa-4b5d-9765-956c8ca0b6e8"
  }
}
2023-03-28 16:55:00.895277-0700 DatastoreAPISample[87093:3235014] [API] Starting network task for query 3D3F733D-5C32-4347-ABFF-7005324B4BB5
success(DatastoreAPISample.Post(id: "895f4988-e7fa-4b5d-9765-956c8ca0b6e8", name: "Post1.updated", blog: nil, createdAt: nil, updatedAt: nil))

```


Can you provide the verbose log when you invoke the custom request? You can enable verbose logs by:

```
Amplify.Logging.logLevel = .verbose
```

**Comment by fr33css** [comment]
@royjit 

Sorry my informations is not clear enough.

```
extension GraphQLRequest {

    static func getPostId(byId id: String) -> GraphQLRequest<Post> {
        let document = """
        query GetPost($id: ID!) {
          getPost(id: $id) {
            id
            title
            userID
            __typename
          }
        }
        """
        return GraphQLRequest<Post>(
            document: document,
            variables: ["id": id],
            responseType: Post.self,
            decodePath: "getPost"
        )
    }
}
```

According the __type is for getting nest datas.
Is it possible to write needed nest data informations?(which meaning i don't extra informations)

```
static func getPostId(byId id: String) -> GraphQLRequest<Post> {
        let document = """
        query GetPost($id: ID!) {
          getPost(id: $id) {
            id
            title
            userID
            __typename
              items {
                comment {
                  id
                  postId
                  content
                  userId
               }
             }
          }
        }
        """
        return GraphQLRequest<Post>(
            document: document,
            variables: ["id": id],
            responseType: Post.self,
            decodePath: "getPost"
        )
    }
```

**Comment by royjit** [comment]
Hey @fr33css are you still facing this issue? Will you be able to provide verbose logs for the query to debug further? Please provide the customer query you are using and the query that you are making. This way we can see if we can reproduce the same behavior.
